web/packages/teleterm/src/ui/utils/retryWithRelogin.ts

```diff
@@ -78,6 +80,51 @@ export async function retryWithRelogin<T>(
   return await actionToRetry();
 }
 
+/**
+ * useRetryWithRelogin is the hook version of retryWithRelogin. Typically used in conjunction with
+ * useAsync and useCallback.
+ *
+ * @example Accepting arguments.
+ * const [attempt, createGateway] = useAsync(async (port: string) =>
+ *   useRetryWithRelogin(
+ *     ctx,
+ *     clusterUri,
+ *     useCallback(async () => {
+ *       try {
+ *         const response = await foo.createGateway(port)
+ *       } catch (error) {
+ *         // Custom error handling logic
+ *       }
+ *     }, [foo])
+ *   )
+ * )
```

Comment:
This example is incorrect - `useRetryWithRelogin` should be called. Otherwise, the return type of `attempt` is `Attempt<() => Promise<void>>`

The correct version should look like this: 
```ts
  const [attempt, createGateway] = useAsync(
    useCallback(
      async (port: string) =>
        useRetryWithRelogin(
          ctx,
          clusterUri,
          useCallback(async () => {
            try {
              const response = await foo.createGateway(port);
            } catch (error) {
              // Custom error handling logic
            }
          }, [port])
        )(),
      [clusterUri, foo, ctx]
    )
  );
```
But it has two disadvantages: we call the hook on our own (I'm not sure if this is correct), and we have another layer of `useCallback`.

I'm wondering if we could add `Args` to the signature of `useRetryWithRelogin`: 
```ts
export function useRetryWithRelogin<Args extends unknown[], T>(
  appContext: IAppContext,
  resourceUri: ClusterOrResourceUri,
  callback: (...args: Args) => Promise<T>
): (...args: Args) => Promise<T> {
```

then we could do this:
```ts
  const [attempt, createGateway] = useAsync(
    useRetryWithRelogin(
      ctx,
      rootClusterUri,
      useCallback(async (port: string) => {
        try {
          const response = await foo.createGateway(port);
        } catch (error) {
          // Custom error handling logic
        }
      }, [foo])
    )
  );
```

I think it should work

Reply:
Ah, thanks for catching that. When I was attempting to write the hook, I had an intuition that it's not going to compose well but after writing out the example with arguments, I thought I had everything figured out.

My intuition was that `useRetryWithRelogin` can be used only in scenarios where the callback passed to `useAsync` accepts no arguments. The callback passed to `retryWithRelogin` cannot accept args – `retryWithRelogin` needs to be able to call the callback again without knowing what the args are. If you use `useAsync` with `retryWithRelogin`, this forces you to capture the arguments of `useAsync` callback in the closure of `retryWithRelogin` callback:

https://github.com/gravitational/teleport/blob/fbba4c2bfada1dbae5f87eaa9ac8289a6e701738/web/packages/teleterm/src/ui/DocumentGateway/useDocumentGateway.ts#L41-L45

And in React world, capturing variables in a closure equals `useCallback` since React needs to know if one of those variables have changed.

So your suggestion won't work either since `retryWithRelogin` doesn't accept args.

I think we can fix the problem in two ways:

1. Explicitly mention that `useRetryWithRelogin` can be used only with thunks. In all other cases you should opt to use regular `retryWithRelogin`.
   * Take a look at that `useDocumentGateway` fragment above. We don't take care of function identity there, but it could be made stable simply by wrapping the callback passed to `useAsync` in `useCallback` and listing deps – everything else would work as usual.
   * However, I worry that having `useRetryWithRelogin` like that would encourage people to use thunks with `useAsync` more which I feel like is a bad practice – if there are some things that can change, you totally should make use of `useCallback` and accept those things as arguments instead.
       * This way the resulting `run` function from `useAsync` can just accept some things as arguments, you don't need to worry about how they affect function identity etc. Thunks destroy that and from my experience lead to incorrect `useEffect` usage.
2. Change `retryWithRelogin` to return a generic function which accepts arguments instead of having `retryWithRelogin` call the callback directly. However, I'm too tired to consider all consequences of this and it'd be too much work for now.
3. Get rid of `useRetryWithRelogin` and just use `retryWithRelogin` + `useCallback` which is what I did in the recent commit.

Reply:
Ah, I forgot that the callback shouldn't take any props.

I agree with your decision, let's use a plain function + `useCallback` for now.